### PR TITLE
Update Intel IGC driver from kernel 5.10.150

### PR DESCRIPTION
**Pull request by andrew64k**
Signed-off-by: Andrew Lindh <andrew@netplex.net>

### CODE REVIEW

File: SPECS/igc-module.spec [hunk 1]
```diff
@@ -1,13 +1,19 @@
+%define vendor_name Intel
+%define vendor_label intel
+%define driver_name igc
```
Comment (stormi) [inline SPECS/igc-module.spec]: Are they required? Citrix adds these in their vendor drivers, but we don't use them in this spec file.

File: SOURCES/0002-backport.patch [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+ };
+ 
+ void igc_up(struct igc_adapter *adapter);
+--- a/igc_base.c	2022-09-23 12:17:01.000000000 +0000
++++ b/igc_base.c	2022-09-24 07:07:24.414598890 +0000
+@@ -205,6 +205,7 @@
+ 	case IGC_DEV_ID_I220_V:
+ 	case IGC_DEV_ID_I225_K:
+ 	case IGC_DEV_ID_I225_K2:
++	case IGC_DEV_ID_I226_K:
```
Comment (stormi) [inline SOURCES/0002-backport.patch]: What is the goal of this addition?

File: SOURCES/0002-backport.patch [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+ 	case IGC_DEV_ID_I225_LMVP:
+ 	case IGC_DEV_ID_I225_IT:
+ 	case IGC_DEV_ID_I226_LM:
+--- a/igc_defines.h	2022-09-23 12:17:01.000000000 +0000
++++ b/igc-module/igc_defines.h	2022-09-24 06:29:34.767001177 +0000
+@@ -160,6 +160,7 @@
+ #define IGC_NVM_RW_REG_START	1    /* Start operation */
+ #define IGC_NVM_RW_ADDR_SHIFT	2    /* Shift to the address bits */
+ #define IGC_NVM_POLL_READ	0    /* Flag for polling for read complete */
++#define IGC_NVM_DEV_STARTER	5    /* Dev_starter Version */
```
Comment (stormi) [inline SOURCES/0002-backport.patch]: What is the goal of this addition?

File: SOURCES/0002-backport.patch [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
++
+ 	strlcpy(drvinfo->bus_info, pci_name(adapter->pdev),
+ 		sizeof(drvinfo->bus_info));
+ 
+@@ -1018,11 +1035,11 @@
+ 	switch (cmd->flow_type) {
+ 	case TCP_V4_FLOW:
+ 		cmd->data |= RXH_L4_B_0_1 | RXH_L4_B_2_3;
+-		fallthrough;
++//		fallthrough;
```
Comment (stormi) [inline SOURCES/0002-backport.patch]: So I think I know why this is removed, but it would be good to mention it in a comment or in the patch description.

File: SOURCES/0002-backport.patch [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+ 	case SCTP_V6_FLOW:
+ 	case AH_ESP_V6_FLOW:
+ 	case AH_V6_FLOW:
+@@ -1899,7 +1916,7 @@
+ }
+ 
+ static const struct ethtool_ops igc_ethtool_ops = {
+-	.supported_coalesce_params = ETHTOOL_COALESCE_USECS,
++//	.supported_coalesce_params = ETHTOOL_COALESCE_USECS,
```
Comment (stormi) [inline SOURCES/0002-backport.patch]: Is this something not supported by kernel 4.19?

File: SOURCES/0002-backport.patch [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
++
++	scnprintf(adapter->fw_version,
++			sizeof(adapter->fw_version),
++			"%x:%x",
++			nvm_version,
++			gphy_version);
++
++	strscpy(drvinfo->fw_version, adapter->fw_version,
++		sizeof(drvinfo->fw_version));
++
```
Comment (stormi) [inline SOURCES/0002-backport.patch]: Why is this needed?

**Review by stormi — COMMENTED** [review]
Thanks for the pull request!

Could you add a comment on top of each patch explaining why it's needed and what changes? Especially the second patch, to document the reasons for the changes, like you would do in a commit message if the patch was a commit rather than a patch. I'm adding my naïve questions in various places of the patch.

File: SPECS/igc-module.spec (hunk 1 above)
Comment (andrew64k) [reply SPECS/igc-module.spec]: No, not required.

File: SOURCES/0002-backport.patch (hunk 2 above)
Comment (andrew64k) [reply SOURCES/0002-backport.patch]: To support the new I226 which is just the I225 v4 with a different ID. I have a new card on the way for empirical testing. Reviewing Intel's code post there is at least one more addition to make.

File: SOURCES/0002-backport.patch (hunk 3 above)
Comment (andrew64k) [reply SOURCES/0002-backport.patch]: To support showing the version from ethtool.

File: SOURCES/0002-backport.patch (hunk 4 above)
Comment (andrew64k) [reply SOURCES/0002-backport.patch]: Old GCC failure. "fallthrough" was added later compilers.

File: SOURCES/0002-backport.patch (hunk 5 above)
Comment (andrew64k) [reply SOURCES/0002-backport.patch]: Not supported in 4.19

File: SOURCES/0002-backport.patch (hunk 6 above)
Comment (andrew64k) [reply SOURCES/0002-backport.patch]: It is for showing version from ethtool It was listed as "to do" in the 5.10 source code. So did it by taking code from a later kernel.

**Comment by andrew64k** [comment]
It's one big diff.

I can split it into a few different parts related to the changes if  
that's important.


-- 
   Andrew Lindh
  ***@***.***
NETPLEX 860-233-1111



Quoting Samuel Verschelde ***@***.***>:

File: SPECS/igc-module.spec [hunk 7]
```diff
@@ -2,12 +2,18 @@
 
 Summary: Driver for igc-module
 Name: igc-module
-Version: 4.20.17
+Version: 5.10.146
 Release: 2%{?dist}
```
Comment (stormi) [inline SPECS/igc-module.spec]: The `Release` tag should be reset to 1 each time the Version is updated.

File: SPECS/igc-module.spec [hunk 8]
```diff
@@ -47,6 +53,12 @@ find %{buildroot}/lib/modules/%{kernel_version} -name "*.ko" -type f | xargs chm
 /lib/modules/%{kernel_version}/*/*.ko
 
 %changelog
+* Sat Oct 01 2022 Andrew Lindh <andrew@netplex.net> - 5.10.146-1
+- Update to kernel source 5.10.146 and split patches into specific functions
+
+* Fri Sep 23 2022 Andrew Lindh <andrew@netplex.net> - 5.10.145-1
+- Update Intel I225 driver from 5.10.145 kernel source
+
```
Comment (stormi) [inline SPECS/igc-module.spec]: As we did not merge nor build the `5.10.145-1` package, only the latest changelog entry is necessary.

(Note: before merging I'll squash all your commits into one, for a cleaner git history on the main branches. So the intermediary 5.10.145-1 will disappear)

**Review by stormi — COMMENTED** [review]
Looks good. Only remaining comments are about the Release tag and the changelog.

**Comment by andrew64k** [comment]
I got my first test Intel I226 pcie card and the old driver does NOT correctly recognize it (as expected).
This new driver version DOES recognize the I226 card and works. The I226 is also informally known as the I225 v4.
The I225 v1 and v2 are know to have issues. The I225 v3 is used by most devices and seems to work well.
It is guessed that I226 will help distance the name from the older I225 versions to make it clear that the updated hardware.

**Comment by stormi** [comment]
My last two comments from 1 week ago are all what's left to be fixed for me to merge the PR.

**Comment by andrew64k** [comment]
I'll update it to reflect a single current dash 1 version and one comment about it being a new build.

I have a new 2.5G switch and new Intel/AMD machines in my test lab so I'll run a new full Host/Guest/NFS/iSCSI/VLAN/MTU/usage/performance test on the driver.

**Comment by hoerup** [comment]
@andrew64k  did you find time to make the last few tweaks ?

**Comment by stormi** [comment]
Same question from me :). We're going to release a public XCP-ng 8.3 alpha soon and including the updated drivers would certainly be appreciated by those who need them.

**Comment by andrew64k** [comment]
I'll do it today.

-- 
   Andrew Lindh
  ***@***.***
NETPLEX 860-233-1111



Quoting Samuel Verschelde ***@***.***>:

File: SPECS/igc-module.spec [hunk 9]
```diff
@@ -53,15 +53,6 @@ find %{buildroot}/lib/modules/%{kernel_version} -name "*.ko" -type f | xargs chm
...
-* Fri Sep 23 2022 Andrew Lindh <andrew@netplex.net> - 5.10.145-1
-- Update Intel I225 driver from 5.10.145 kernel source
-
-* Fri Sep 16 2022 Samuel Verschelde <stormi-xcp@ylix.fr> - 4.20.17-2
-- Rebuild for XCP-ng 8.3
-
-* Tue Feb 15 2022 Rushikesh Jadhav <rushikesh7@gmail.com> - 4.20.17-1
-- Added driver igc from kernel 4.20.17
+* Thu Oct 27 2022 Andrew Lindh <andrew@netplex.net> - 5.10.150-1
+- Use kernel source 5.10.150 and patches for XCP driver support
```
Comment (stormi) [inline SPECS/igc-module.spec:57]: This should not remove the changelog entries for 4.20.17-1 and 4.20.17-2 as they were already in the file before this PR.

**Comment by stormi** [comment]
I checked the contents of `igc-module-5.10.150.tar.gz` against the sources of kernel 5.10.150: identical as expected.

Squashing and merging. Thanks for this contribution and don't hesitate to open new PRs to update the driver when needed!

**Comment by stormi** [comment]
Koji build task: http://koji.xcp-ng.org/taskinfo?taskID=26445

**Comment by andrew64k** [comment]
The IGC code looks rather stable for the last several 5.10 LTS kernel updates. I know there will be (or should be)  some updates to the upstream code at some date for the new I226. I'll check for future updates and also address any XCP code issues I can. It looks to be a more popular on-board chip for new systems.

**Comment by stormi** [comment]
I added igc-module as a dependency to the vendor-drivers RPM, so it will be installed by default. Regarding `install.img`, did you have to do anything specific, or was installing the RPM into it and repacking it enough?

**Comment by andrew64k** [comment]
I added it to the ISO boot kernel and install.img file system as a loadable module at boot so the system would have networking available for the install (like other normal ethernet interfaces).

I don't know if the XCP installer can automatically load new extra packages at the start to support additional features during the install (not just a required package installed on the new XCP system).

**Comment by stormi** [comment]
I've done it the past. It should be enough to install it inside the initrd (install.img) for the kernel to load if when needed.
